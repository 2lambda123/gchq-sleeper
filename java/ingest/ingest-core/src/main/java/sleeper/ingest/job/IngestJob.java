/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.ingest.job;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

public class IngestJob {
    private final String tableName;
    private final String id;
    private final List<String> files;

    private IngestJob(Builder builder) {
        tableName = builder.tableName;
        id = builder.id;
        files = builder.files;
    }

    public static Builder builder() {
        return new Builder();
    }

    public List<String> getValidationFailures() {
        List<String> validationFailures = new ArrayList<>();
        if (files == null) {
            validationFailures.add("Missing property \"files\"");
        } else if (files.contains(null)) {
            validationFailures.add("One of the files was null");
        }
        if (tableName == null) {
            validationFailures.add("Missing property \"tableName\"");
        }
        return validationFailures;
    }

    public String getTableName() {
        return tableName;
    }

    public String getId() {
        return id;
    }

    public List<String> getFiles() {
        return files;
    }

    public int getFileCount() {
        return Optional.ofNullable(files).map(List::size).orElse(0);
    }

    public Builder toBuilder() {
        return builder().id(id).files(files).tableName(tableName);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        IngestJob ingestJob = (IngestJob) o;
        return Objects.equals(id, ingestJob.id) &&
                Objects.equals(tableName, ingestJob.tableName) &&
                Objects.equals(files, ingestJob.files);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, tableName, files);
    }

    @Override
    public String toString() {
        return "IngestJob{" +
                "id='" + id + '\'' +
                ", tableName='" + tableName + '\'' +
                ", files=" + files +
                '}';
    }

    public static final class Builder {
        private String tableName;
        private String id;
        private List<String> files;

        private Builder() {
        }

        public Builder tableName(String tableName) {
            this.tableName = tableName;
            return this;
        }

        public Builder id(String id) {
            this.id = id;
            return this;
        }

        public Builder files(List<String> files) {
            this.files = files;
            return this;
        }

        public Builder files(String... files) {
            return files(Arrays.asList(files));
        }

        public IngestJob build() {
            return new IngestJob(this);
        }
    }
}
