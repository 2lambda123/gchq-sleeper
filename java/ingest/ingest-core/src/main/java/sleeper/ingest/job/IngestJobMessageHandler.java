/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.ingest.job;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.ingest.job.status.IngestJobStatusStore;
import sleeper.ingest.job.status.IngestJobValidatedEvent;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import static sleeper.ingest.job.status.IngestJobValidatedEvent.ingestJobRejected;

public class IngestJobMessageHandler<T> {
    private static final Logger LOGGER = LoggerFactory.getLogger(IngestJobMessageHandler.class);
    private final IngestJobStatusStore ingestJobStatusStore;
    private final Function<String, T> deserialiser;
    private final Function<T, IngestJob> toIngestJob;
    private final BiFunction<T, IngestJob, T> applyIngestJobChanges;
    private final Function<List<String>, List<String>> expandDirectories;
    private final Supplier<String> jobIdSupplier;
    private final Supplier<Instant> timeSupplier;

    private IngestJobMessageHandler(Builder<T> builder) {
        ingestJobStatusStore = builder.ingestJobStatusStore;
        deserialiser = builder.deserialiser;
        toIngestJob = builder.toIngestJob;
        applyIngestJobChanges = builder.applyIngestJobChanges;
        expandDirectories = builder.expandDirectories;
        jobIdSupplier = builder.jobIdSupplier;
        timeSupplier = builder.timeSupplier;
    }

    public static Builder<IngestJob> forIngestJob() {
        return builder()
                .deserialiser(new IngestJobSerDe()::fromJson)
                .toIngestJob(job -> job)
                .applyIngestJobChanges((job, changedJob) -> changedJob);
    }

    public static Builder<?> builder() {
        return new Builder<>();
    }

    public Optional<T> deserialiseAndValidate(String message) {
        T job;
        try {
            job = deserialiser.apply(message);
            LOGGER.info("Deserialised message to ingest job {}", job);
        } catch (RuntimeException e) {
            LOGGER.warn("Deserialisation failed for message: {}", message, e);
            ingestJobStatusStore.jobValidated(
                    ingestJobRejected(jobIdSupplier.get(), message, timeSupplier.get(),
                            "Error parsing JSON. Reason: " + Optional.ofNullable(e.getCause()).orElse(e).getMessage()));
            return Optional.empty();
        }
        IngestJob ingestJob = toIngestJob.apply(job);
        IngestJob.Builder ingestJobBuilder = ingestJob.toBuilder();
        String jobId = ingestJob.getId();
        if (jobId == null || jobId.isBlank()) {
            jobId = jobIdSupplier.get();
            LOGGER.info("Null or blank id provided. Generated new id: {}", jobId);
            ingestJobBuilder.id(jobId);
        }

        List<String> files = ingestJob.getFiles();
        List<String> validationFailures = new ArrayList<>();
        if (files == null) {
            validationFailures.add("Missing property \"files\"");
        } else if (files.contains(null)) {
            validationFailures.add("One of the files was null");
        }
        if (ingestJob.getTableName() == null) {
            validationFailures.add("Missing property \"tableName\"");
        }
        if (!validationFailures.isEmpty()) {
            LOGGER.warn("Validation failed: {}", validationFailures);
            ingestJobStatusStore.jobValidated(
                    refusedEventBuilder()
                            .jobId(jobId)
                            .tableName(ingestJob.getTableName())
                            .jsonMessage(message)
                            .reasons(validationFailures.stream()
                                    .map(failure -> "Model validation failed. " + failure)
                                    .collect(Collectors.toUnmodifiableList()))
                            .build());
            return Optional.empty();
        }

        List<String> expandedFiles = expandDirectories.apply(files);
        if (expandedFiles.isEmpty()) {
            LOGGER.warn("Could not find one or more files for job: {}", job);
            ingestJobStatusStore.jobValidated(
                    ingestJobRejected(jobId, message, timeSupplier.get(), "Could not find one or more files"));
            return Optional.empty();
        }

        LOGGER.info("No validation failures found");
        return Optional.of(applyIngestJobChanges.apply(job, ingestJob.toBuilder().id(jobId).files(expandedFiles).build()));
    }

    private IngestJobValidatedEvent.Builder refusedEventBuilder() {
        return IngestJobValidatedEvent.builder()
                .validationTime(timeSupplier.get());
    }

    public static final class Builder<T> {
        private IngestJobStatusStore ingestJobStatusStore;
        private Function<String, T> deserialiser;
        private Function<T, IngestJob> toIngestJob;
        private BiFunction<T, IngestJob, T> applyIngestJobChanges;
        private Function<List<String>, List<String>> expandDirectories;
        private Supplier<String> jobIdSupplier = () -> UUID.randomUUID().toString();
        private Supplier<Instant> timeSupplier = Instant::now;

        private Builder() {
        }

        public Builder<T> ingestJobStatusStore(IngestJobStatusStore ingestJobStatusStore) {
            this.ingestJobStatusStore = ingestJobStatusStore;
            return this;
        }

        public <N> Builder<N> deserialiser(Function<String, N> deserialiser) {
            this.deserialiser = (Function<String, T>) deserialiser;
            return (Builder<N>) this;
        }

        public Builder<T> toIngestJob(Function<T, IngestJob> toIngestJob) {
            this.toIngestJob = toIngestJob;
            return this;
        }

        public Builder<T> applyIngestJobChanges(BiFunction<T, IngestJob, T> applyIngestJobChanges) {
            this.applyIngestJobChanges = applyIngestJobChanges;
            return this;
        }

        public Builder<T> expandDirectories(Function<List<String>, List<String>> expandDirectories) {
            this.expandDirectories = expandDirectories;
            return this;
        }

        public Builder<T> jobIdSupplier(Supplier<String> jobIdSupplier) {
            this.jobIdSupplier = jobIdSupplier;
            return this;
        }

        public Builder<T> timeSupplier(Supplier<Instant> timeSupplier) {
            this.timeSupplier = timeSupplier;
            return this;
        }

        public IngestJobMessageHandler<T> build() {
            return new IngestJobMessageHandler<>(this);
        }
    }
}
