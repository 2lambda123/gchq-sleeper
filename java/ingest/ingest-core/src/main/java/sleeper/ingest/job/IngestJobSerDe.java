/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.ingest.job;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import org.apache.hadoop.conf.Configuration;

import sleeper.configuration.properties.InstanceProperties;

import java.util.Optional;

import static sleeper.utils.HadoopPathUtils.expandDirectories;

public class IngestJobSerDe {
    private final Gson gson;
    private final Gson gsonPrettyPrinting;
    private final Configuration configuration;
    private final InstanceProperties properties;

    public IngestJobSerDe() {
        this(new Configuration(), new InstanceProperties());
    }

    public IngestJobSerDe(Configuration configuration, InstanceProperties properties) {
        this.gson = new GsonBuilder().create();
        this.gsonPrettyPrinting = new GsonBuilder()
                .setPrettyPrinting()
                .create();
        this.configuration = configuration;
        this.properties = properties;
    }

    public String toJson(IngestJob ingestJob) {
        return gson.toJson(ingestJob);
    }

    public String toJson(IngestJob ingestJob, boolean prettyPrint) {
        if (prettyPrint) {
            return gsonPrettyPrinting.toJson(ingestJob);
        }
        return toJson(ingestJob);
    }

    public IngestJob fromJson(String jsonIngestJob) {
        return gson.fromJson(jsonIngestJob, IngestJob.class);
    }

    public Optional<IngestJob> fromJsonExpandingDirs(String jsonIngestJob) {
        IngestJob ingestJob = fromJson(jsonIngestJob);
        return expandDirectories(ingestJob.getFiles(), configuration, properties)
                .map(files -> ingestJob.toBuilder().files(files).build());
    }
}
