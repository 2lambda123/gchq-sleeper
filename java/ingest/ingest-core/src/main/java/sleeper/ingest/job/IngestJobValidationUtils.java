/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.ingest.job;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.ingest.job.status.IngestJobStatusStore;

import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import static sleeper.ingest.job.status.IngestJobValidatedEvent.ingestJobRejected;

public class IngestJobValidationUtils {
    private static final Logger LOGGER = LoggerFactory.getLogger(IngestJobValidationUtils.class);

    private IngestJobValidationUtils() {
    }

    public static <T> Optional<T> deserialiseAndValidate(
            String message, Function<String, T> deserialiser, Function<T, List<String>> getValidationFailures,
            IngestJobStatusStore ingestJobStatusStore, Supplier<String> invalidJobIdSupplier, Supplier<Instant> timeSupplier) {
        T job;
        try {
            job = deserialiser.apply(message);
            LOGGER.info("Deserialised message to ingest job {}", job);
        } catch (RuntimeException e) {
            LOGGER.warn("Deserialisation failed: {}", e.getStackTrace());
            ingestJobStatusStore.jobValidated(
                    ingestJobRejected(invalidJobIdSupplier.get(), message, timeSupplier.get(),
                            "Error parsing JSON. Reason: " + Optional.ofNullable(e.getCause()).orElse(e).getMessage()));
            return Optional.empty();
        }
        List<String> validationFailures = getValidationFailures.apply(job);
        if (validationFailures.isEmpty()) {
            LOGGER.info("No validation failures found");
            return Optional.of(job);
        } else {
            LOGGER.warn("Validation failed: {}", validationFailures);
            ingestJobStatusStore.jobValidated(
                    ingestJobRejected(invalidJobIdSupplier.get(), message, timeSupplier.get(),
                            validationFailures.stream()
                                    .map(failure -> "Model validation failed. " + failure)
                                    .collect(Collectors.toList())));
            return Optional.empty();
        }
    }
}
