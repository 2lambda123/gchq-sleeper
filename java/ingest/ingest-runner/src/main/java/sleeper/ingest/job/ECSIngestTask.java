/*
 * Copyright 2022 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.ingest.job;

import com.amazonaws.services.cloudwatch.AmazonCloudWatch;
import com.amazonaws.services.cloudwatch.AmazonCloudWatchClientBuilder;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClientBuilder;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.sqs.AmazonSQS;
import com.amazonaws.services.sqs.AmazonSQSClientBuilder;
import org.apache.hadoop.conf.Configuration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import sleeper.configuration.jars.ObjectFactory;
import sleeper.configuration.jars.ObjectFactoryException;
import sleeper.configuration.properties.InstanceProperties;
import sleeper.configuration.properties.table.TablePropertiesProvider;
import sleeper.core.iterator.IteratorException;
import sleeper.ingest.status.store.task.DynamoDBIngestTaskStatusStore;
import sleeper.ingest.task.IngestTask;
import sleeper.ingest.task.IngestTaskStatusStore;
import sleeper.statestore.StateStoreException;
import sleeper.statestore.StateStoreProvider;
import sleeper.utils.HadoopConfigurationProvider;
import software.amazon.awssdk.services.s3.S3AsyncClient;

import java.io.IOException;
import java.util.UUID;

import static sleeper.configuration.properties.UserDefinedInstanceProperty.S3A_INPUT_FADVISE;

public class ECSIngestTask {
    private ECSIngestTask() {
    }

    private static final Logger LOGGER = LoggerFactory.getLogger(ECSIngestTask.class);

    public static void main(String[] args) throws IOException, StateStoreException, IteratorException, ObjectFactoryException {
        if (1 != args.length) {
            System.err.println("Error: must have 1 argument (s3Bucket)");
            System.exit(1);
        }

        long startTime = System.currentTimeMillis();
        AmazonDynamoDB dynamoDBClient = AmazonDynamoDBClientBuilder.defaultClient();
        AmazonSQS sqsClient = AmazonSQSClientBuilder.defaultClient();
        AmazonCloudWatch cloudWatchClient = AmazonCloudWatchClientBuilder.defaultClient();
        AmazonS3 s3Client = AmazonS3ClientBuilder.defaultClient();

        String s3Bucket = args[0];
        InstanceProperties instanceProperties = new InstanceProperties();
        instanceProperties.loadFromS3(s3Client, s3Bucket);

        ObjectFactory objectFactory = new ObjectFactory(instanceProperties, s3Client, "/tmp");
        String localDir = "/mnt/scratch";
        String taskId = UUID.randomUUID().toString();

        IngestTask ingestTask = createIngestTask(objectFactory, instanceProperties, localDir,
                taskId, s3Client, dynamoDBClient, sqsClient, cloudWatchClient, S3AsyncClient.create(),
                ingestHadoopConfiguration(instanceProperties));
        ingestTask.run();

        s3Client.shutdown();
        LOGGER.info("Shut down s3Client");
        sqsClient.shutdown();
        LOGGER.info("Shut down sqsClient");
        dynamoDBClient.shutdown();
        LOGGER.info("Shut down dynamoDBClient");
        long finishTime = System.currentTimeMillis();
        double runTimeInSeconds = (finishTime - startTime) / 1000.0;
        LOGGER.info("IngestFromIngestJobsQueueRunner total run time = {}", runTimeInSeconds);
    }

    public static IngestTask createIngestTask(ObjectFactory objectFactory,
                                              InstanceProperties instanceProperties,
                                              String localDir,
                                              String taskId,
                                              AmazonS3 s3Client,
                                              AmazonDynamoDB dynamoDBClient,
                                              AmazonSQS sqsClient,
                                              AmazonCloudWatch cloudWatchClient,
                                              S3AsyncClient s3AsyncClient,
                                              Configuration hadoopConfiguration) {
        TablePropertiesProvider tablePropertiesProvider = new TablePropertiesProvider(s3Client, instanceProperties);
        StateStoreProvider stateStoreProvider = new StateStoreProvider(dynamoDBClient, instanceProperties, hadoopConfiguration);
        IngestTaskStatusStore taskStore = DynamoDBIngestTaskStatusStore.from(dynamoDBClient, instanceProperties);
        IngestJobRunner ingestJobRunner = new IngestJobRunner(
                objectFactory,
                instanceProperties,
                tablePropertiesProvider,
                stateStoreProvider,
                localDir,
                s3AsyncClient,
                hadoopConfiguration);
        IngestJobQueueConsumer queueConsumer = new IngestJobQueueConsumer(sqsClient, cloudWatchClient, instanceProperties);
        return new IngestTask(
                queueConsumer, taskId, taskStore, ingestJobRunner::ingest);
    }

    private static Configuration ingestHadoopConfiguration(InstanceProperties instanceProperties) {
        Configuration conf = HadoopConfigurationProvider.getConfigurationForECS(instanceProperties);
        conf.set("fs.s3a.experimental.input.fadvise", instanceProperties.get(S3A_INPUT_FADVISE));
        return conf;
    }
}
