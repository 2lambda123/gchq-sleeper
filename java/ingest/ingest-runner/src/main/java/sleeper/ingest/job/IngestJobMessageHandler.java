/*
 * Copyright 2022-2023 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.ingest.job;

import org.apache.hadoop.conf.Configuration;

import sleeper.configuration.properties.instance.InstanceProperties;
import sleeper.ingest.job.status.IngestJobStatusStore;

import java.time.Instant;
import java.util.Optional;
import java.util.UUID;
import java.util.function.Supplier;

public class IngestJobMessageHandler {
    private final IngestJobValidationUtils ingestJobValidationUtils;

    public IngestJobMessageHandler(Configuration configuration, InstanceProperties instanceProperties,
                                   IngestJobStatusStore ingestJobStatusStore) {
        this(configuration, instanceProperties, ingestJobStatusStore, () -> UUID.randomUUID().toString(), Instant::now);

    }

    public IngestJobMessageHandler(Configuration configuration, InstanceProperties instanceProperties,
                                   IngestJobStatusStore ingestJobStatusStore,
                                   Supplier<String> invalidJobIdSupplier, Supplier<Instant> timeSupplier) {
        this.ingestJobValidationUtils = new IngestJobValidationUtils(ingestJobStatusStore, invalidJobIdSupplier,
                timeSupplier, configuration, instanceProperties);
    }

    public Optional<IngestJob> handleMessage(String message) {
        return ingestJobValidationUtils.deserialiseAndValidate(message,
                        new IngestJobSerDe()::fromJson, IngestJob::getValidationFailures)
                .flatMap(job -> ingestJobValidationUtils.expandDirectories(job,
                        files -> job.toBuilder().files(files).build()));
    }
}
