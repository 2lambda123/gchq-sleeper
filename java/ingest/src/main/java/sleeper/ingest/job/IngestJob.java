/*
 * Copyright 2022 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.ingest.job;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class IngestJob {
    private final String tableName;
    private final String id;
    private final List<String> files;

    public IngestJob(String tableName, String id, List<String> files) {
        this.tableName = tableName;
        this.id = id;
        this.files = files;
    }

    private IngestJob(Builder builder) {
        tableName = builder.tableName;
        id = builder.id;
        files = builder.files;
    }

    public static Builder builder() {
        return new Builder();
    }

    public String getTableName() {
        return tableName;
    }

    public String getId() {
        return id;
    }

    public List<String> getFiles() {
        return files;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        IngestJob ingestJob = (IngestJob) o;
        return Objects.equals(id, ingestJob.id) &&
                Objects.equals(tableName, ingestJob.tableName) &&
                Objects.equals(files, ingestJob.files);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, tableName, files);
    }

    @Override
    public String toString() {
        return "IngestJob{" +
                "id='" + id + '\'' +
                ", tableName=" + tableName +
                ", files=" + files +
                '}';
    }

    public static final class Builder {
        private String tableName;
        private String id;
        private List<String> files;

        private Builder() {
        }

        public Builder tableName(String tableName) {
            this.tableName = tableName;
            return this;
        }

        public Builder id(String id) {
            this.id = id;
            return this;
        }

        public Builder files(List<String> files) {
            this.files = files;
            return this;
        }

        public Builder files(String... files) {
            return files(Arrays.asList(files));
        }

        public IngestJob build() {
            return new IngestJob(this);
        }
    }
}
