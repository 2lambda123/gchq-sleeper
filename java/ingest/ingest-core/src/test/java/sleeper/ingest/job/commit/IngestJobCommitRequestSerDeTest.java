/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.ingest.job.commit;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import sleeper.core.record.process.RecordsProcessed;
import sleeper.core.record.process.RecordsProcessedSummary;
import sleeper.core.statestore.FileReference;
import sleeper.ingest.job.IngestJob;

import java.time.Duration;
import java.time.Instant;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class IngestJobCommitRequestSerDeTest {

    private final IngestJobCommitRequestSerDe serDe = new IngestJobCommitRequestSerDe();

    @Test
    void shouldSerialiseIngestJobCommitRequest() throws Exception {
        // Given
        IngestJob job = IngestJob.builder()
                .id("test-job-id")
                .files(List.of("file1.parquet", "file2.parquet"))
                .tableId("test-table-id")
                .tableName("test-table-name")
                .build();
        FileReference file1 = FileReference.builder()
                .filename("file1.parquet")
                .partitionId("root")
                .numberOfRecords(100L)
                .onlyContainsDataForThisPartition(true)
                .build();
        FileReference file2 = FileReference.builder()
                .filename("file2.parquet")
                .partitionId("root")
                .numberOfRecords(200L)
                .onlyContainsDataForThisPartition(true)
                .build();
        IngestJobCommitRequest commit = new IngestJobCommitRequest(job, "test-task", List.of(file1, file2),
                new RecordsProcessedSummary(
                        new RecordsProcessed(300, 300),
                        Instant.parse("2024-05-01T10:58:00Z"), Duration.ofMinutes(1)));

        // When
        String json = serDe.toJsonPrettyPrint(commit);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(commit);
        Approvals.verify(json);
    }
}
