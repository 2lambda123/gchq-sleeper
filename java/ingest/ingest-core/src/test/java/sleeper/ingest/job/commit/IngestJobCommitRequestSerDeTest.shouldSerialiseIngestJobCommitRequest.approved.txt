{
  "type": "IngestJobCommitRequest",
  "request": {
    "ingestJob": {
      "id": "test-job-id",
      "tableName": "test-table-name",
      "tableId": "test-table-id",
      "files": [
        "file1.parquet",
        "file2.parquet"
      ]
    },
    "taskId": "test-task",
    "fileReferenceList": [
      {
        "filename": "file1.parquet",
        "partitionId": "root",
        "numberOfRecords": 100,
        "jobId": null,
        "lastStateStoreUpdateTime": null,
        "countApproximate": false,
        "onlyContainsDataForThisPartition": true
      },
      {
        "filename": "file2.parquet",
        "partitionId": "root",
        "numberOfRecords": 200,
        "jobId": null,
        "lastStateStoreUpdateTime": null,
        "countApproximate": false,
        "onlyContainsDataForThisPartition": true
      }
    ],
    "startTime": 1714561080000,
    "finishTime": 1714561140000,
    "recordsRead": 300,
    "recordsWritten": 300
  }
}